### ISSUE TITLE
Revisit renaming with qualified access

### ISSUE THREAD

**Issue by lydia-duncan** [chapel-lang/chapel#14738]
Currently, you can only use the old name when performing qualified naming, and the new name for unqualified naming.  For example:

```chapel
module L {
    var x: int = 42;
}

use L only x as y;
writeln(x);   // <-- Error today
writeln(y);   // L.x, works
writeln(L.x); // L.x, works
writeln(L.y); // <-- Error today, under discussion
```

The thinking was that the module itself was not changing, so the rename isn't visible in that scope.

However, when revisiting, we think that using the new name should be allowed.  E.g.

```chapel
module L {
    var x: int = 42;
}

use L only x as y;
writeln(L.y); // L.x, works
```

The current proposal is to continue to allow qualified access with the old name as well - we do think it is a bit odd that the old name would still be available in this form, but we also think that it won't cause issues if it is still available.  Forbidding access with the old name will have a greater potential to break backwards compatibility - however, it seems more likely for a user to *either* use qualified access *or* rename a symbol, rather than doing both for the same symbol, so doing so probably wouldn't be the end of the world.

Otherwise, the only potential for breaking backwards compatibility happens if a symbol is renamed to the same name as a symbol that wasn't included, e.g.

```chapel
module L {
   var x: int;
   var y: int;
}
module User {
   use L only x as y;

   proc main() {
      writeln(L.y); // If this follows the renaming scheme, there is no way to
      // access the original L.y
   }
}
```

In this case, the original L.y will be inaccessible to the client code in any form.  However, we think such cases are unlikely to be a problem in practice.

This discussion was spawned off of behavior noticed when implementing #10799

**Comment by BryantLam** [comment]
I'm not sure what to do about this behavior. It's both consistent and not consistent at the same time, likely because of Chapel's syntax choice for `use`ing module items. I'm emphatically a fan of Rust's syntax choice; [I'll demonstrate with an example in Rust.](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ccb76ba89f9796f65ace6138ba8cb63a)

I recommend looking at that link for additional comments. The general way that Rust's `use` statement works is that it imports the referred-to item. If that item is a module, it imports the module symbol (qualified access). If that item is a variable in a module, it imports that variable (unqualified access). By allowing `use` on more than just module symbols, it lets the user directly control exactly what symbols get imported into the local scope and from where.

With that said, let's pretend the following Rust code is in a ___different file___ (unlike the link I have above) so that we avoid the "modules are in scope if they're in the same file" rule.
```rust
// Rust: M.rs
mod M {
    pub const x: i32 = 1;
    pub const y: i32 = 2;
}
```

### Block 1: Only qualified access
```rust
// Rust
use M;
println!("{} {}", M::x, M::y);
```
The equivalent Chapel statement: `import M;`

### Block 2: Only unqualified access
```rust
// Rust
use M::{x, y};
println!("{} {}", x, y);
```
Chapel has no equivalent statement.

### Block 2a: Qualified + Unqualified access
```rust
// Rust
use M; // <-- This line imports item `M`, like Block 1
use M::{x, y};
println!("{} {}", x, y);
```
The equivalent Chapel statement: `use M only x, y`.

Note that in Rust, you can also import all symbols in a module for unqualified access like so: `use M::*;` The Chapel equivalent would then be `use M;`.

### Block 3: Rename variables without conflict
```rust
// Rust
use M;
use M::{x as y};
println!("{} {}", y, M::y);
```

The case for this issue. This Rust code does exactly what it says it does with no ambiguity because of the syntax choice that Rust uses.

### Block 3a: Rename module and variable
```rust
// Rust
use M as L;
use M::{x as y};
println!("{} {} {}", L::x, L::y, y); // 1 2 1
```
More complicated example of Block 3.

--

Now, returning to Block 3, what should Chapel do? I think this problem is made harder by the fact that Chapel `use` statements have three behaviors #13978:

1. unqualified access to a module (note: not *item* like in Rust)
2. qualified access to a module // This is the problem.
3. if `public use M` in current scope and current scope is a module definition, then a dependent scope can inherit behaviors 1 and 2 of `M` through another `use` statement of the currently being defined module

In my ideal world, I would prefer if the `import` statement worked on items rather than module symbols (items include module symbols).
- Qualified access only: `import M`
- Unqualified access only: `import M.*`
- Rename a variable: `import M.x as y` or `import M.{x as y}`
  - Imports only the variable, not the full module; in fact, you shouldn't be allowed to refer to `M` at all if you don't `import M`.

I think designing renaming and re-exports for `import` statements is more important. Then, we should just make `use` statements be some combination of `import` statements for the sake of consistency:

```chapel
// Chapel
module L {
    var x: int = 42;
}

// What does this do?
// use L only x as y;

// Equivalent would be:
import L;
import L.{x as y};
writeln(x);   // Illegal: no x in scope
writeln(y);   // Ok
writeln(L.x); // Ok
writeln(L.y); // Illegal: no L.y in L
```

In other words, I think no changes need to happen today if we decompose the `use` statement into its `import` equivalents.

**Comment by BryantLam** [comment]
(Sorry if you spent too much time on my wild goose chase! I think `use` statements cannot be special and should just be some combination of `import` statements in order to make sense of what this feature should do.)

**Comment by BryantLam** [comment]
To be a contrarian, the design of import statements is important since a combination of them could result in a different interpretation of what renaming-on-`use` actually does.

For example,
```chapel
// What does this do?
// use L only x as y;

// Maybe another equivalent:
import L with x as y; // <-- notional syntax that changes L for some reason
import L.{x as y};
writeln(x);   // Illegal: no x in scope
writeln(y);   // Ok
writeln(L.x); // Ok??
writeln(L.y); // Ok, unless conflict??
```

In retrospect, I can't think of a good case of where a user would want to rename/change the contents of a module.

> however, it seems more likely for a user to _either_ use qualified access _or_ rename a symbol, rather than doing both for the same symbol, so doing so probably wouldn't be the end of the world.

It is as you say, this might not be a real problem to solve. That said, my initial confusion and surprise means that renaming-on-use is yet another feature (quirk?) that has to be learned. But if Chapel expects that `use` statements will be replaced with `import` statements in the long term, then I don't know how much these quirks matter other than to make them consistent in some definition with `import`.

**Comment by lydia-duncan** [comment]
I can get behind not making this decision until we've finalized the `import`/`use` combination world.

I'll make some import variation proposals and push on that

**Comment by lydia-duncan** [comment]
In looking at Rust's system, I think your summary of Block 2a, 3, and 3a are not accurate.  Or rather, not complete.  It looks like there's multiple ways to write qualified access - `use M;` or `use M::{self}`.  The latter allows you to specify other symbols at the same time as allowing access to the module and qualified access.  With that in mind, Block 2a could also be written:
```Rust
use M::{x, y, self};
println!("{} {} {} {}", x, y, M::x, M::y);
```

Block 3 could also be written:
```Rust
use M::{x as y, self};
println!("{} {}", y, M::y);
```
because Rust doesn't override the explicit naming version with the new name when they're in the same use statement.
However, that doesn't avoid confusion in the same way as your other version does.  But you can similarly accomplish the explicitness of your other version in Chapel by writing:
```chapel
// Chapel
use M only;
use M only x as y;
writeln(y);
writeln(M.y);
```
We just don't have a way to say "give me only unqualified access" in Chapel today, so the first use statement is redundant with the second one present.

Similarly, Block 3a can be written as:
```Rust
// Rust
use M::{x as y, self as L};
println!("{} {} {}", L::x, L::y, y); // 1 2 1
```
